cx/slack/server.js

```diff
@@ -0,0 +1,124 @@
+
+const slackEvents = createEventAdapter(slackSigningSecret);
+const slackClient = new WebClient(slackToken);
+
+// Imports the Google Cloud Some API library
+const {SessionsClient} = require('@google-cloud/dialogflow-cx');
+/**
+ * Example for regional endpoint:
+ *   const locationId = 'us-central1'
+ *   const client = new SessionsClient({apiEndpoint: 'us-central1-dialogflow.googleapis.com'})
+ */
+const client = new SessionsClient({apiEndpoint: locationId + '-dialogflow.googleapis.com'});
+
+/**
+ * Converts Slack request to a detectIntent request.
+ */
+function slackToDetectIntent(slackRequest,sessionPath){
+    const request = {
+        session: sessionPath,
+        queryInput: {
+            text: {
+                text: slackRequest.text,
+            },
+            languageCode,
+        },
+    };
+
+    return request;
+}
+
+/**
+ * Converts detectIntent response to a Slack message request. 
+ */
+function detectIntentToSlackMessage(response,id){
+    var agentResponse = '';
+    
+    for (const message of response.queryResult.responseMessages) {
+        if (message.text) {
+            agentResponse += `${message.text.text}\n`;
+        };
+    };
+    
+    if(agentResponse.length != ''){
+        const request = {
+            channel: id,
+            text: agentResponse
+        };
+        return request;
+    };
+};
+
+/**
+ * Takes as input a request from Slack and converts the request to
+ * detectIntent request which is used to call the detectIntent() function
+ * and finally output the response given by detectIntent().
+ */
+async function detectIntentResponse(slackRequest) {
+    const sessionId = slackRequest.user;
+    const sessionPath = client.projectLocationAgentSessionPath(
+        projectId,
+        locationId,
+        agentId,
+        sessionId
+    );
+    console.info(sessionPath);
+    
+    request = slackToDetectIntent(slackRequest, sessionPath);
+    const [response] = await client.detectIntent(request);
+
+    return response;
+};
+
+/**
+ * Checks if the request is coming from a bot and if it is not 
+ * it will call detectIntentResponse() with the request as the input
+ * and sends the response to the Slack server. The initial check to make
+ * sure the request is not from a bot is there to prevent the Integration
+ * from replying to its own messages.
+ */
+slackEvents.on('message', (event) => {
+    if(event.bot_id == '' || event.bot_id == null){
+        (async () => {
+            const bot = await slackClient.auth.test();
+            const response = await detectIntentResponse(event);
+            if(event.channel_type == 'im'){
+                const request = detectIntentToSlackMessage(response, event.user);
+                try {
+                    await slackClient.chat.postMessage(request)
+                } catch (error) {
+                    console.log(error.data)
+                }
+            }else if(event.text.includes(bot.user_id)){
```

Comment:
I found that the check is necessary, otherwise the integration reponds to every message in the channals it is in. I put message.groups and message.channals as they are also events that the ES integration is subscibed to.

Reply:
It looks more clean and reliable to just handle the `app_mention`, rather than check the bot ID in the message text.

It might make sense to not subscribe to `message.groups` or `message.channels`, if the bot only wants to handle the message mentioning it.

Reply:
I tried app_mention before but the app_mention only subscribes to mentions and not other messages like ims. I also tried removing message.groups and message.channals and trying the integration but it seems to disable reading or posting messages to channals when I deleted those subscriptions. The integration needs to also respond to instant messages to the Slack app and also through app mentions.

Reply:
I tried to propose subscribes to the following events:
* `app_mention`
* `message.im`

IMs are covered by `message.im`. The messages from channels are covered by `app_mention`.

For the permission to read/post message, it should be controlled by `oauth_config.scopes`.

Reply:
This integration actually doesn't use app_mention. I tried using app_mention but it made it so that I am unable to both use instant messaging and read and write to channels at the same time. That is why I have to check is the text message has the bot id in it before posting a message in the channal. On the other hand, I can remove app_mention since I am not really using it.

Reply:
It might be able to work if `oauth_config.scopes` is set up correctly.

E.g., for the bot to read the mention message, it probably need the [app_mentions:read](https://api.slack.com/scopes/app_mentions:read) scope as indicated in the spec https://api.slack.com/events/app_mention.

Reply:
I took your suggestion and looked at the oAuth permission, and I found permission that were unnecessary and after I removed them the integration is now able to use app_mention. I have updated server.js, server_test.js and the app manifest to reflect the new changes and to  make use of app_mention.

Reply:
Cool. Good to know that it works.